traffic_ops/traffic_ops_golang/deliveryservice/deleteoldcerts.go

```diff
@@ -0,0 +1,136 @@
+package deliveryservice
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/apache/trafficcontrol/lib/go-log"
+	"github.com/apache/trafficcontrol/lib/go-tc"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/config"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/dbhelpers"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/riaksvc"
+
+	"github.com/basho/riak-go-client"
+)
+
+// DeleteOldCerts asychronously deletes HTTPS certificates in Riak which have no corresponding delivery service in the database.
+//
+// Note the delivery service may still be in the CRConfig! Therefore, this should only be called immediately after a CRConfig Snapshot.
+//
+// This creates a goroutine, and immediately returns. It returns an error if there was an error preparing the delete routine, such as an error creating a db transaction.
+//
+// Note because it is asynchronous, this may return a nil error, but the asynchronous goroutine may error when fetching or deleting the certificates. If such an error occurs, it will be logged to the error log.
+//
+// If certificate deletion is already being processed by a goroutine, this immediately returns nil.
```

Comment:
q: why? What if there is a DS deletion + snapshot shortly after this snapshot while an existing transaction is still in progress. Wouldn't we want that 2nd attempt to simply wait until the current transaction is done being processed, then start its own transaction to delete certs for the latest deleted DS?

Reply:
Right. But we don't want them to get backed up. If people do 50 snapshots at once, then the first deletion job starts, we don't want to do 49 more, we just want to do the 1 more to fix the "race."

The ideal solution is to have a queue of 1, and if 1 is going, and 49 more come in, the first one waits, and the other 48 say "someone is already waiting, we're done."

That's just a little more complicated, and it seemed like an unlikely race. It'd only be different if someone deleted a DS and snapshotted, all in-between the time the job got DSes from Riak and deleted them.

I'll try to see how much more work that "queue of 1" is, though.

Reply:
Yeah I agree on it being fairly unlikely, but I figure if it's not too much trouble to avoid it then maybe we should. We could also just push that bit out into a separate PR for later if it's nontrivial.

Reply:
I just realized, it's more important than that (and more difficult), because it has to be per-cdn.

Right now, if snapshots of 2 CDNs come in at once, it'll only run the first. That's a problem.

I'll try to have it done by the end of the day.

Reply:
Oh! Good catch. That seems like a much more likely race.

Reply:
Done